File: 37-sudoku-solver/sudoku-solver.py
class Solution:
    def solveSudoku(self, board: List[List[str]]) -> None:
        """
        Do not return anything, modify board in-place instead.
        """
        row = [set() for _ in range(9)]
        col = [set() for _ in range(9)]
        box = [set() for _ in range(9)]
        for i in range(9):
            for j in range(9):
                if board[i][j] != ".":
                    num = board[i][j]
                    row[i].add(num)
                    col[j].add(num)
                    box[(i // 3) * 3 + j // 3].add(num)
                         
        def backtrack(board,i=0,j=0):
            if i == 9 :
                return True
            if board[i][j] != ".":
                if j<8:
                    return backtrack(board,i,j+1)
                else:
                    return backtrack(board,i+1,0)
            else:
                for k in range(1,10):
                    c= str(k)
                    if c not in row[i] and c not in col[j] and c not in box[(i // 3) * 3 + j // 3]:
                        board[i][j] = str(k)
                        row[i].add(c)
                        col[j].add(c)
                        box[(i // 3) * 3 + j // 3].add(c)
                        if j<8:
                            if backtrack(board,i,j+1):
                                return True
                        else:
                            if backtrack(board,i+1,0):
                                return True
                        board[i][j] = "."
                        row[i].remove(c)
                        col[j].remove(c)
                        box[(i // 3) * 3 + j // 3].remove(c)
            return False
        backtrack(board)
                     
            
        

File: 37-sudoku-solver/sudoku-solver.cpp
class Solution {
public:
    void solveSudoku(vector<vector<char>>& board) {
       vector<vector<bool>> row(9,vector<bool>(9,false));
       vector<vector<bool>> col(9,vector<bool>(9,false));
       vector<vector<bool>> box(9,vector<bool>(9,false));
       for(int i=0;i<9;i++) {
        for(int j=0;j<9;j++) {
            if (board[i][j]!='.') {
                int num = board[i][j] - '1';
                row[i][num] = true;
                col[j][num] = true;
                box[(i/3)*3+(j/3)][num] = true;
            }
        }
       }
       backtrack(board,0,0,row,col,box);
    }
    bool backtrack(vector<vector<char>>& board,int i,int j,vector<vector<bool>>& row,vector<vector<bool>>& col,vector<vector<bool>>& box)
    {
        if (i==9) {
            return true;
        }
        int ni,nj;
        if (j==8) {
            ni = i+1;
            nj = 0;
        } else {
            ni = i;
            nj = j+1;
        }
        
        if (board[i][j] != '.') {
            return backtrack(board,ni,nj,row,col,box);
        }
        for (int k=1;k<=9;k++) {
            int b = (i/3) *3 + (j/3);
            if (!row[i][k-1] && !col[j][k-1] && !box[b][k-1]) {
                board[i][j] = (char)(k+'0');
                row[i][k-1] = col[j][k-1] = box[b][k-1] = true;
                if (backtrack(board,ni,nj,row,col,box)) return true;
                row[i][k-1] = col[j][k-1] = box[b][k-1] = false;
                board[i][j] = '.';
            }
        }
        return false;
    }
};